velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -117,15 +208,52 @@ void HiveDataSink::createWriter(
   options.schema = inputType_;
   // Without explicitly setting flush policy, the default memory based flush
   // policy is used.
-  auto writerParameters = getWriterParameters(partitionName);
-
-  writerInfo_.emplace_back(std::make_shared<HiveWriterInfo>(*writerParameters));
 
+  std::optional<std::string> partitionName = partitionChannels_.size() > 0
+      ? std::optional<std::string>(makePartitionName(
+            input_,
+            partitionChannels_,
+            partitionRows_[partitionId]->as<vector_size_t>()[0]))
+      : std::nullopt;
+  auto writerParameters = getWriterParameters(partitionName);
   auto writePath = fs::path(writerParameters->writeDirectory()) /
       writerParameters->writeFileName();
+
   auto sink = dwio::common::DataSink::create(writePath);
-  writers_.push_back(std::make_unique<Writer>(
-      options, std::move(sink), *connectorQueryCtx_->memoryPool()));
+  writers_[partitionId] = std::make_unique<Writer>(
+      options, std::move(sink), *connectorQueryCtx_->memoryPool());
+  writerInfo_[partitionId] =
+      std::make_shared<HiveWriterInfo>(*writerParameters);
+}
+
+void HiveDataSink::computePartitionRowCountsAndIndices(
+    uint64_t numPartitionIds,
+    memory::MemoryPool* pool) {
+  partitionSizes_.resize(numPartitionIds);
+  std::fill(partitionSizes_.begin(), partitionSizes_.end(), 0);
+  partitionRows_.resize(numPartitionIds);
+  std::fill(partitionRows_.begin(), partitionRows_.end(), nullptr);
+
+  std::vector<vector_size_t*> rawIndexValues(numPartitionIds);
+  uint64_t numRows = partitionIds_.size();
+
+  for (vector_size_t row = 0; row < numRows; row++) {
+    uint64_t id = partitionIds_[row];
+
+    if (partitionRows_[id] == nullptr) {
+      partitionRows_[id] = allocateIndices(numRows, pool);
```

Comment:
It would be nice to skip populating this buffer when writing to a single partition. PartitionIdGenerator should be able to tell us whether there is only one partition.

Reply:
Yeah, revised to take a bool for single partition and skip populating the buffer if the bool is true.

Reply:
Hi @mbasmanova Masha, here I need to figure out a way to tell if it is writing to a single partition.

You suggest we should be able to get this info from PartitionIdGenerator, and I get it. The way I'm trying here is to keep track of the recentMaxPartitionId_ and recentMinPartitionId_ inside PartitionIdGenerator, and compare them to see if there's only a single partition id. Inside PartitionIdGenerator we traverse the rows of input anyway, so keeping track of max & min is not adding more traversal overhead. Also we need to expose recentMaxPartitionId.

https://github.com/facebookincubator/velox/blob/d3accd7b4066ab7b6eafef1e2e25e553ef8805c5/velox/connectors/hive/PartitionIdGenerator.cpp#L83-L84

Also in your previous PR, you have this comment,
https://github.com/facebookincubator/velox/blob/4653ec4329b4366f392c5519e1ea3728b337e1e3/velox/connectors/hive/PartitionIdGenerator.cpp#L61-L63

I think here we need to tell if the current input writes to a single partition, while looking at every VectorHasher's # of UniqueValues gives us if all inputs in history write to a single partition, which is more strict.

Let me know what you think.